File: riveredge-frontend/src/services/businessConfig.ts
/**
 * 业务配置 API 服务
 *
 * 变更说明（2026 重构）：
 * - 业务蓝图设置已下线；功能开关改由菜单管理控制，审核改由流程设置控制。
 * - 本文件仅保留纯「参数」读写（bom / sales / purchase / finance 等）
 *   以及 PRO 功能检查相关接口。旧版 modules/nodes/running-mode/
 *   complexity-presets/templates 接口已删除。
 */

import { apiRequest } from './api';

export interface BusinessConfig {
  parameters: Record<string, Record<string, any>>;
}

export interface ProcessParameterUpdateRequest {
  category: string;
  parameter_key: string;
  value: any;
}

export interface BatchProcessParameterUpdateRequest {
  parameters: Record<string, Record<string, any>>;
}

export interface BusinessConfigSchema {
  processRegistry?: Record<string, string[]>;
  processRegistryMeta?: Record<string, { labelKey?: string; descriptionKey?: string }>;
  processRegistryParamMeta?: Record<string, Record<string, { labelKey?: string; descriptionKey?: string }>>;
  processRegistryControlMeta?: Record<
    string,
    Record<string, { type?: 'boolean' | 'number' | 'string' | 'color' | 'select'; min?: number; max?: number; options?: { value: string; labelKey: string }[] }>
  >;
  parameterRegistry?: Record<string, string[]>;
  parameterRegistryMeta?: Record<string, { labelKey?: string; descriptionKey?: string }>;
  parameterRegistryParamMeta?: Record<string, Record<string, { labelKey?: string; descriptionKey?: string }>>;
  parameterRegistryControlMeta?: Record<
    string,
    Record<string, { type?: 'boolean' | 'number' | 'string' | 'color' | 'select'; min?: number; max?: number; options?: { value: string; labelKey: string }[] }>
  >;
  parameterKeys: Record<string, string[]>;
  parameterImplementation?: Record<string, Record<string, boolean>>;
}

export async function getBusinessConfigSchema(): Promise<BusinessConfigSchema> {
  return apiRequest<BusinessConfigSchema>('/infra/business-config/schema', {
    method: 'GET',
  });
}

export async function getBusinessConfig(): Promise<BusinessConfig> {
  return apiRequest<BusinessConfig>('/infra/business-config', {
    method: 'GET',
  });
}

/** 是否开启试运营模式（读取 parameters.common.trial_run_mode，默认 false） */
export function isTrialRunModeEnabled(config: BusinessConfig | null | undefined): boolean {
  return Boolean(config?.parameters?.common?.trial_run_mode);
}

/** 详情抽屉全链路模式：关闭 / 启用（含节点时间） / 只显示单据（不展示节点创建时间） */
export type DetailFullChainMode = 'off' | 'on' | 'documents_only';

export function resolveDetailFullChainMode(
  config: BusinessConfig | null | undefined,
): DetailFullChainMode {
  const common = config?.parameters?.common;
  const raw = common?.detail_full_chain_mode ?? common?.detail_full_chain_enabled;
  if (typeof raw === 'boolean') return raw ? 'documents_only' : 'off';
  if (raw === 'off' || raw === 'on' || raw === 'documents_only') return raw;
  return 'documents_only';
}

/** 启用模式下展示节点创建时间 */
export function isDetailFullChainShowCreatedAt(
  config: BusinessConfig | null | undefined,
): boolean {
  return resolveDetailFullChainMode(config) === 'on';
}

/** 详情抽屉是否展示全链路跟踪 Tab（mode !== off；默认展示） */
export function isDetailFullChainEnabled(config: BusinessConfig | null | undefined): boolean {
  return resolveDetailFullChainMode(config) !== 'off';
}

/** 详情抽屉是否展示操作记录（默认 true） */
export function isDetailOperationLogEnabled(config: BusinessConfig | null | undefined): boolean {
  const raw = config?.parameters?.common?.detail_operation_log_enabled;
  return raw === undefined ? true : Boolean(raw);
}

/** 详情抽屉基本信息是否展示更新时间（默认 true） */
export function isDetailBasicUpdatedAtEnabled(config: BusinessConfig | null | undefined): boolean {
  const raw = config?.parameters?.common?.detail_basic_updated_at_enabled;
  return raw === undefined ? true : Boolean(raw);
}

/** 详情抽屉按字段隐藏的时间项：key 为 `{documentType}.{dataIndex}`，true 表示隐藏 */
export function resolveDetailTimeFieldHiddenMap(
  config: BusinessConfig | null | undefined,
): Record<string, boolean> {
  const raw = config?.parameters?.common?.detail_time_field_hidden;
  if (!raw || typeof raw !== 'object' || Array.isArray(raw)) return {};
  const out: Record<string, boolean> = {};
  for (const [key, value] of Object.entries(raw as Record<string, unknown>)) {
    if (typeof key === 'string' && key.trim()) out[key] = Boolean(value);
  }
  return out;
}

/** 与当前库字段 decimal_places 对齐的配置上限（配置页 max 同源） */
export const NUMERIC_PRECISION_STORAGE_CEILING = {
  quantity: 2,
  price: 4,
  amount: 2,
} as const;

export type NumericPrecisionKind = keyof typeof NUMERIC_PRECISION_STORAGE_CEILING;

export type NumericPrecisionSettings = {
  quantity: number;
  price: number;
  amount: number;
};

function clampDecimalPlaces(raw: unknown, defaultPlaces: number, maxPlaces: number): number {
  const n = Number(raw);
  if (!Number.isFinite(n)) return defaultPlaces;
  return Math.max(0, Math.min(maxPlaces, Math.trunc(n)));
}

/** 从业务配置解析数值精度（默认 2；超出库字段上限则截断） */
export function resolveNumericPrecisionFromConfig(
  config: BusinessConfig | null | undefined,
  kind: NumericPrecisionKind,
): number {
  const common = config?.parameters?.common;
  const key =
    kind === 'quantity'
      ? 'quantity_decimal_places'
      : kind === 'price'
        ? 'price_decimal_places'
        : 'amount_decimal_places';
  return clampDecimalPlaces(common?.[key], 2, NUMERIC_PRECISION_STORAGE_CEILING[kind]);
}

/** 按配置小数位格式化数字（固定位，含尾零） */
export function formatByNumericPrecision(
  value: number | string | null | undefined,
  places: number,
  fallback = '-',
): string {
  if (value === null || value === undefined || value === '') return fallback;
  const n = Number(value);
  if (!Number.isFinite(n)) return fallback;
  const safePlaces = Math.max(0, Math.min(6, Math.trunc(places)));
  return n.toLocaleString('zh-CN', {
    minimumFractionDigits: safePlaces,
    maximumFractionDigits: safePlaces,
  });
}

export async function updateProcessParameter(
  request: ProcessParameterUpdateRequest
): Promise<{ success: boolean; message: string; category: string; parameter_key: string; value: any }> {
  return apiRequest('/infra/business-config/parameters/update', {
    method: 'POST',
    data: request,
  });
}

export async function batchUpdateProcessParameters(
  request: BatchProcessParameterUpdateRequest
): Promise<{ success: boolean; message: string; updated_count: number }> {
  return apiRequest('/infra/business-config/parameters/batch-update', {
    method: 'POST',
    data: request,
  });
}

export interface ProFeatureAccessCheck {
  has_access: boolean;
  is_pro_feature: boolean;
  current_plan: string;
  upgrade_message?: string;
}

export interface ProFeaturesList {
  has_pro_access: boolean;
  current_plan: string;
  pro_modules: string[];
  pro_parameters: Record<string, string[]>;
}

export interface AuditRequiredMapResponse {
  audit_required: Record<string, boolean>;
}

export async function getAuditRequiredMap(): Promise<Record<string, boolean>> {
  const res = await apiRequest<AuditRequiredMapResponse>('/infra/business-config/audit-required', {
    method: 'GET',
  });
  return res?.audit_required ?? {};
}

export async function checkProFeatureAccess(
  featureType: string,
  featureCode: string
): Promise<ProFeatureAccessCheck> {
  return apiRequest<ProFeatureAccessCheck>(
    `/infra/business-config/pro-features/check?feature_type=${featureType}&feature_code=${featureCode}`,
    { method: 'GET' }
  );
}

export async function getProFeaturesList(): Promise<ProFeaturesList> {
  return apiRequest<ProFeaturesList>('/infra/business-config/pro-features/list', {
    method: 'GET',
  });
}

File: riveredge-backend/src/infra/api/business_config/business_config.py
"""
业务配置 API 模块

提供业务「参数」配置相关的 API 接口（如 BOM 多版本、销售/采购/财务参数等）。

变更说明（2026 重构）：
- 业务蓝图设置已下线；功能是否开启由「菜单管理」决定，是否审核由「流程设置（ApprovalProcess）」决定。
- 本模块不再提供 nodes/modules/running-mode/complexity-presets/templates 等蓝图相关接口，
  仅保留纯参数 (`parameters.*`) 的读取与更新，供系统参数 Tab 使用。
- PRO 功能列表/校验接口仍然保留，与套餐相关。

Author: Luigi Lu
"""

from fastapi import APIRouter, HTTPException, Depends
from loguru import logger

from infra.schemas.business_config import (
    BusinessConfigResponse,
    ProcessParameterUpdateRequest,
    BatchProcessParameterUpdateRequest,
)
from infra.services.business_config_service import BusinessConfigService
from infra.api.deps.deps import get_current_user
from core.api.deps.deps import get_current_tenant
from infra.models.user import User
from infra.exceptions.exceptions import ValidationError, NotFoundError, BusinessLogicError

router = APIRouter(prefix="/business-config", tags=["Business Config"])


@router.get("/schema", summary="获取业务配置 schema")
async def get_config_schema(
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    获取业务配置 schema（参数键）

    返回可配置的参数分类及键，供前端动态渲染系统参数表单。
    蓝图相关 (processRegistry / allNodes 等) 已移除。
    """
    from infra.services.business_config_service import (
        _build_parameter_keys_schema,
        _build_process_registry_schema,
        _build_process_registry_meta_schema,
        _build_process_registry_param_meta_schema,
        _build_process_registry_control_meta_schema,
        _build_parameter_registry_schema,
        _build_parameter_registry_meta_schema,
        _build_parameter_registry_param_meta_schema,
        _build_parameter_registry_control_meta_schema,
        _build_parameter_implementation_schema,
    )
    # 说明：processRegistry 与 parameterRegistry 都属于系统「参数」分组，
    # 前端配置中心据此渲染「流程设置」「参数设置」两个参数 Tab；
    # 蓝图节点相关 (allNodes 等) 已在蓝图下线时移除。
    return {
        "processRegistry": _build_process_registry_schema(),
        "processRegistryMeta": _build_process_registry_meta_schema(),
        "processRegistryParamMeta": _build_process_registry_param_meta_schema(),
        "processRegistryControlMeta": _build_process_registry_control_meta_schema(),
        "parameterRegistry": _build_parameter_registry_schema(),
        "parameterRegistryMeta": _build_parameter_registry_meta_schema(),
        "parameterRegistryParamMeta": _build_parameter_registry_param_meta_schema(),
        "parameterRegistryControlMeta": _build_parameter_registry_control_meta_schema(),
        "parameterKeys": _build_parameter_keys_schema(),
        "parameterImplementation": _build_parameter_implementation_schema(),
    }


@router.get("", response_model=BusinessConfigResponse, summary="获取业务配置")
async def get_business_config(
    current_user: User = Depends(get_current_user),
    tenant_id: int = Depends(get_current_tenant),
) -> BusinessConfigResponse:
    """
    获取业务配置（仅 parameters 有效；nodes/modules 字段已废弃，读取时返回空或历史值）。
    """
    try:
        config = await BusinessConfigService().get_business_config(tenant_id)
        return BusinessConfigResponse(**config)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error(f"获取业务配置失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取业务配置失败: {str(e)}")


@router.get("/audit-required", summary="批量获取单据审核开关")
async def get_audit_required_map(
    current_user: User = Depends(get_current_user),
    tenant_id: int = Depends(get_current_tenant),
) -> dict:
    """
    返回当前组织各单据节点是否需要审核。
    审核真源为 ApprovalProcess(code=node_key, is_active=true)。
    """
    try:
        result = await BusinessConfigService().get_audit_required_map(tenant_id)
        return {"audit_required": result}
    except Exception as e:
        logger.error(f"获取审核开关失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取审核开关失败: {str(e)}")


@router.post("/parameters/update", summary="更新流程参数")
async def update_process_parameter(
    request: ProcessParameterUpdateRequest,
    current_user: User = Depends(get_current_user),
    tenant_id: int = Depends(get_current_tenant),
) -> dict:
    """更新指定分类下的单个参数。"""
    try:
        return await BusinessConfigService().update_process_parameter(
            tenant_id=tenant_id,
            category=request.category,
            parameter_key=request.parameter_key,
            value=request.value,
        )
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"更新流程参数失败: {e}")
        raise HTTPException(status_code=500, detail=f"更新流程参数失败: {str(e)}")


@router.post("/parameters/batch-update", summary="批量更新流程参数")
async def batch_update_process_parameters(
    request: BatchProcessParameterUpdateRequest,
    current_user: User = Depends(get_current_user),
    tenant_id: int = Depends(get_current_tenant),
) -> dict:
    """批量更新多个分类的参数。"""
    try:
        return await BusinessConfigService().batch_update_process_parameters(
            tenant_id=tenant_id,
            parameters=request.parameters,
        )
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except ValidationError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"批量更新流程参数失败: {e}")
        raise HTTPException(status_code=500, detail=f"批量更新流程参数失败: {str(e)}")


@router.get("/pro-features/check", summary="检查 PRO 版功能访问权限")
async def check_pro_feature_access(
    feature_type: str,
    feature_code: str,
    current_user: User = Depends(get_current_user),
    tenant_id: int = Depends(get_current_tenant),
) -> dict:
    try:
        return await BusinessConfigService().check_pro_feature_access(
            tenant_id=tenant_id,
            feature_type=feature_type,
            feature_code=feature_code,
        )
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except BusinessLogicError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except Exception as e:
        logger.error(f"检查 PRO 版功能访问权限失败: {e}")
        raise HTTPException(status_code=500, detail=f"检查 PRO 版功能访问权限失败: {str(e)}")


@router.get("/pro-features/list", summary="获取 PRO 版功能列表")
async def get_pro_features_list(
    current_user: User = Depends(get_current_user),
    tenant_id: int = Depends(get_current_tenant),
) -> dict:
    try:
        return await BusinessConfigService().get_pro_features_list(tenant_id)
    except NotFoundError as e:
        raise HTTPException(status_code=404, detail=str(e))
    except Exception as e:
        logger.error(f"获取 PRO 版功能列表失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取 PRO 版功能列表失败: {str(e)}")
